src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComActivator.cs

```diff
@@ -177,6 +197,72 @@ private static Type FindClassType(Guid clsid, string assemblyName, string typeNa
             throw new COMException(string.Empty, CLASS_E_CLASSNOTAVAILABLE);
         }
 
+        private static AssemblyLoadContext GetALC(string assemblyPath)
+        {
+            AssemblyLoadContext alc;
+            return LazyInitializer.EnsureInitialized(ref s_AssemblyLoadContexts).TryGetValue(assemblyPath, out alc) ?
+                alc:
+                AddNewALC(assemblyPath);
+
+            AssemblyLoadContext AddNewALC(string path)
+            {
+                // The static collection of ALCs is readonly. If a new ALC entry needs to be added,
```

Comment:
I would just take a lock here instead of trying to build lock-free algorithms.

Reply:
I can do the lock thing if the spurious ALCs creation from above is really that costly.